traits/observers/_trait_event_notifier.py

```diff
@@ -0,0 +1,207 @@
+    to the user's handler.
+
+    A ``TraitEventNotifier`` keeps a reference count in order to address
+    situations where a same object is repeated inside a container
+    and one would not want to fire the same change handler multiple times
+    (see enthought/traits#237). For that purpose, a notifier keeps track of
+    the ``HasTraits`` instance (called ``target``) on which the user applies
+    the observers, keeps a reference count internally, and it also needs to
+    determine whether another notifier refers to the same change handler and
+    ``HasTraits`` instance.
+
+    Since there is only one reference count associated with a notifier,
+    each notifier is expected to be added to only one observable.
+    """
+
+    def __init__(
+            self, *, handler, target,
+            event_factory, prevent_event, dispatcher):
+        """
+        Parameters
+        ----------
+        handler : callable(event)
+            The user's handler to receive the change event.
+            The event object is returned by the ``event_factory``.
+            If the handler is an instance method, then a weak reference is
+            created for the method. If the instance is garbage collected,
+            the notifier will be muted.
+        target : object
+            An object for defining the context of the notifier.
+            A weak reference is created for the target.
+            If the target is garbage collected, the notifier will be muted.
+            This target is typically an instance of ``HasTraits`` and will be
+            seen by the user as the "owner" of the change handler.
+            This is also used for distinguishing one notifier from another
+            notifier wrapping the same handler.
+        event_factory : callable(*args, **kwargs) -> object
+            A factory function for creating the event object to be sent to
+            the handler. The call signature must be compatible with the
+            call signature expected by the observable this notifier is used
+            with. e.g. for CTrait, the call signature will be
+            ``(object, name, old, new)``.
+        prevent_event : callable(event) -> boolean
+            A callable for controlling whether the user handler should be
+            invoked. It receives the event created by the event factory and
+            returns true if the event should be prevented, false if the event
+            should be fired.
+        dispatcher : callable(handler, event)
+            A callable for dispatching the handler, e.g. on a different
+            thread or on a GUI event loop. ``event`` is the object
+            created by the event factory.
+        """
+        # This is such that the notifier does not prevent
+        # the target from being garbage collected.
+        self.target = weakref.ref(target)
+
+        if isinstance(handler, types.MethodType):
+            self.handler = weakref.WeakMethod(handler)
```

Comment:
Previously there was a callback in both weakref for removing this notifier from the observable when either the handler or the target is deleted. That requires keeping a reference to the list of notifiers in the observable. If we don't do the cleanup, we may have some silenced notifiers lying around if the code that puts up the notifiers don't do a good job of taking them down (e.g. traitsui).  This may be a bit inefficient, but I am not sure how common that will be.

Reply:
19 days after making this comment: I am now more in favour of not having a weakref callback for removing the notifier. Without the weakref callback, there is only one way to remove the notifier: `remove_from`. Currently `remove_from` raises an exception if the notifier to be removed is not found. This is very helpful for catching programming error. If the code is right, every call to `add_to` should be undone by a call to `remove_from`, like open bracket and close bracket. If we had a weakref callback to remove the notifier from the observable, this introduces a second way for notifiers to be removed, from a random thread, and `remove_from` can no longer raise an exception if the notifier cannot be found.